eni: refactor(metrics) IPAM related metrics

wu0407 opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->
1. fix `cilium_operator_ipam_available` is zero when all node are no need ip.
2. refactor ipam allocation and release metrics.
  2.1 add `cilium_operator_allocation_duration_seconds` and `cilium_operator_release_duration_seconds`
3. rename metrics name
    3.1 `cilium_operator_ipam_available` to `cilium_operator_ipam_available_interfaces`
    3.2 `cilium_operator_ipam_allocation_ops` to `cilium_operator_ipam_ip_allocation_ops`
    3.3 `cilium_operator_ipam_release_ops` to `cilium_operator_ipam_ip_release_ops`

Fixes: #20704
Fixes: #20384

```release-note
Add new ENI IPAM metrics for allocation, release
```

gandro commented on pkg/aws/eni/node.go:
```diff
@@ -528,6 +536,23 @@ func (n *Node) ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Ent
+			if availableOnENI > 0 {
+				remainAvailableENIsCount++
+			}
+
```
This block is copied over directly from `PrepareIPAllocation`, correct? Can we move that logic into a separate function somehow? Otherwise we run into the risk that if one place is changed, that the other one will not, and thus lead to inconsistencies.

wu0407 replied:
resloved in https://github.com/cilium/cilium/compare/ffe5a341ef6bd7ae923aaec4bcd6bac2cc24bfa1..2ef0193d8fb3a85b517ad6d4704022f94016fed6

gandro commented on pkg/azure/ipam/node.go:
```diff
@@ -173,8 +173,43 @@ func (n *Node) ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Ent
+	})
+	if err != nil {
+		return nil, -1, err
+	}
```
Similarly to the comment above, let's move this into a separate function. Otherwise the logic will diverge over time.

wu0407 replied:
Is mean error  handle need into a separate function? I move available logic in separate function.
``` go
availableOnInterface, available := isAvailableInterface(requiredIfaceName, iface, scopedLog)
		if !available {
			return nil
		}
```

gandro replied:
Apologies, I missed that. This is good as is. Minor nit: `remainAvailableENIsCount` is not a good variable name here, since Azure network interfaces are not ENIs. Maybe use  `remainingAvailableInterfaceCount` instead?

gandro commented on pkg/ipam/node.go:
```diff
@@ -353,19 +363,18 @@ func (n *Node) InstanceID() (id string) {
+	// dependent on caller not using the resource after this call.
+	resource = resource.DeepCopy()
+	n.ops.UpdatedNode(resource)
+
```
Could you add a comment to the source as to why this has been moved down? It's not clear to me from context.

Also, I'm nervous about the fact that we now write `n.resource = resource` without creating a copy first. That could introduce data races. Can you keep the deep copy at the top of the function, and only move the `UpdatedNode` call?

wu0407 replied:
this my mistake,  I add comment for that, update n.resource before n.ops.UpdatedNode.

gandro reviewed (changes requested):
Great improvements, thank you! A few minor things which I believe should be addressed before we can merge this.

christarazi reviewed (commented):
Thanks the PR, +1 to Tom's feedback about splitting commits.

tommyp1ckles commented on pkg/azure/ipam/node.go:
```diff
@@ -201,3 +206,27 @@ func (n *Node) GetUsedIPWithPrefixes() int {
+			scopedLog.WithFields(logrus.Fields{
+				"ifaceName":    iface.Name,
+				"requiredName": requiredIfaceName,
+			}).Debug("Not considering interface is available since it does not match the required name")
```
```suggestion
			}).Debug("Not considering interface as available since it does not match the required name")
```

tommyp1ckles commented on pkg/azure/ipam/node.go:
```diff
@@ -201,3 +206,27 @@ func (n *Node) GetUsedIPWithPrefixes() int {
+	scopedLog.WithFields(logrus.Fields{
+		"id":           iface.ID,
+		"numAddresses": len(iface.Addresses),
+	}).Debug("Considering interface is available")
```
```suggestion
	}).Debug("Considering interface as available")
```

tommyp1ckles reviewed (approved):
Azure changes look fine, thanks @wu0407

gandro commented on pkg/ipam/node.go:
```diff
@@ -390,7 +396,7 @@ func (n *Node) recalculate() {
 	scopedLog := n.logger()
 
-	a, err := n.ops.ResyncInterfacesAndIPs(context.TODO(), scopedLog)
+	a, remainAvailableENIsCount, err := n.ops.ResyncInterfacesAndIPs(context.TODO(), scopedLog)
```
Similarly as above, let's call this ` remainingAvailableInterfaceCount` to keep it generic

gandro commented on pkg/ipam/node_manager.go:
```diff
@@ -58,7 +58,7 @@ type NodeOperations interface {
 	// sparingly as this information is kept in sync based on the success
 	// of the functions AllocateIPs(), ReleaseIPs() and CreateInterface().
-	ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Entry) (ipamTypes.AllocationMap, error)
+	ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Entry) (allocatable ipamTypes.AllocationMap, remainAvailableENIsCount int, err error)
```
Similarly as above, let's call this remainingAvailableInterfaceCount to keep it generic

gandro commented on pkg/ipam/node.go:
```diff
@@ -357,15 +357,21 @@ func (n *Node) UpdatedResource(resource *v2.CiliumNode) bool {
-
+	// Update n.resource before n.ops.UpdatedNode. This is to increase the chance
+	// perform a complete n.recalculate() when the nodeManager performs resync
+	// and the operator is startup.
```
nit, better grammar:

```suggestion
	// of performing a complete n.recalculate() execution when the nodeManager performs a resync
	// and the operator is starting up.
```

gandro commented on pkg/ipam/node.go:
```diff
@@ -357,15 +357,21 @@ func (n *Node) UpdatedResource(resource *v2.CiliumNode) bool {
+	// and the operator is startup.
+	// This best effort to solve the metrics `cilium_operator_ipam_available_interfaces`
+	// only contains part of nodes at operator startup. It will be set to the correct
+	// value after the next period nodeManager performs resync.
```
```suggestion
	// This best effort to solve the issue where the metrics `cilium_operator_ipam_available_interfaces`
	// only contains a part of the nodes at operator startup. It will be set to the correct
	// value after the next period where nodeManager performs a resync.
```

gandro reviewed (commented):
Thanks for addressing the feedback, some more nits

gandro commented on pkg/aws/eni/node.go:
```diff
@@ -743,3 +743,20 @@ func (n *Node) GetUsedIPWithPrefixes() int {
+	if n.node.Ops().IsPrefixDelegated() {
+		effectiveLimits = effectiveLimits * option.ENIPDBlockSizeIPv4
+	}
+	return
```
nit:

```suggestion
	return effectiveLimits
```

gandro commented on pkg/azure/ipam/node.go:
```diff
@@ -201,3 +206,27 @@ func (n *Node) GetUsedIPWithPrefixes() int {
 	return len(n.k8sObj.Status.IPAM.Used)
 }
+
+// isAvailableInterface return whether interface is available and the number of available IPs to allocate in interface
```
nit:

```suggestion
// isAvailableInterface returns whether interface is available and the number of available IPs to allocate in interface
```

gandro commented on pkg/ipam/node.go:
```diff
@@ -357,15 +357,21 @@ func (n *Node) UpdatedResource(resource *v2.CiliumNode) bool {
+	// Update n.resource before n.ops.UpdatedNode. This is to increase the chance
+	// of performing a complete n.recalculate() execution when the nodeManager performs a resync
+	// and the operator is starting up.
+	// This best effort to solve the issue where the metrics `cilium_operator_ipam_available_interfaces`
```
nit:

```suggestion
	// This is best effort to solve the issue where the metrics `cilium_operator_ipam_available_interfaces`
```

gandro reviewed (approved):
Thanks, this looks good to me. Three more minor non-blocking nits.

I will be out of office for the next two weeks, so I'll rely on the other reviewers on this PR to drive this to completion.

qmonnet reviewed (approved):
Looks good, thanks!

christarazi commented on pkg/alibabacloud/eni/node.go:
```diff
@@ -172,9 +172,13 @@ func (n *Node) CreateInterface(ctx context.Context, allocation *ipam.AllocationA
+func (n *Node) ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Entry) (available ipamTypes.AllocationMap, remainAvailableENIsCount int, err error) {
+	limits, limitsAvailable := n.getLimits()
+	if !limitsAvailable {
+		return nil, -1, fmt.Errorf("Unable to determine limits")
```
Error strings should be lower-cased.

christarazi commented on pkg/ipam/node_manager.go:
```diff
@@ -58,7 +58,7 @@ type NodeOperations interface {
 	// sparingly as this information is kept in sync based on the success
 	// of the functions AllocateIPs(), ReleaseIPs() and CreateInterface().
-	ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Entry) (ipamTypes.AllocationMap, error)
+	ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Entry) (allocatable ipamTypes.AllocationMap, remainingAvailableInterfaceCount int, err error)
```
I think instead of having named returns here, could you document what the return values are meant to be in the godoc.

wu0407 replied:
it done

christarazi replied:
Could you remove the named returns in the interface declaration? There's no need for them there, just the godoc changes are fine.

wu0407 replied:
Ok, i remove the named returns

christarazi commented on pkg/aws/eni/node.go:
```diff
@@ -507,11 +500,15 @@ func (n *Node) CreateInterface(ctx context.Context, allocation *ipam.AllocationA
+func (n *Node) ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Entry) (available ipamTypes.AllocationMap, remainAvailableENIsCount int, err error) {
+	limits, limitsAvailable := n.getLimits()
+	if !limitsAvailable {
+		return nil, -1, fmt.Errorf("Unable to determine limits")
```
Still seeing error strings as upper-cased. Should be lower-cased

wu0407 replied:
replace all the "Unable to determine limits" to lower

wu0407 committed:
fix cilium_operator_ipam_available is zero when all node are no need ip

This is not fix cilium_operator_ipam_available does not contains all nodes
interface when operator startup.
It will be set to the correct value after the next period nodeManager performs a resync.

Fixes: #20384

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>

joestringer commented on Documentation/operations/upgrade.rst:
```diff
@@ -451,11 +451,20 @@ Helm Options
+* ``cilium_operator_ipam_available`` is removed. Please use ``cilium_operator_ipam_available_interfaces`` instead.
+* ``cilium_operator_ipam_allocation_ops`` is removed. Please use ``cilium_operator_ipam_ip_allocation_ops`` instead.
+* ``cilium_operator_ipam_release_ops`` is removed. Please use ``cilium_operator_ipam_ip_release_ops`` instead.
+* The label of ``status`` in ``cilium_operator_ipam_interface_creation_ops`` is removed.
```
nit: Can you please move this up into a new section for v1.13 upgrade notes? It doesn't make sense to state this in the v1.11 section. If there is not the existing `ADded metrics` titles etc. in the v1.13 section above, please create it.

wu0407 replied:
current no `1.13 Upgrade Notes` in [Documentation/operations/upgrade.rst](https://github.com/cilium/cilium/pull/20755/files/73a4efa0a2fb4eb2d60aba8e2d533955e38423f7#diff-76c7d3d61bd3c6afeb8e936896c7a01046d530840abf499fd118e88e75d1a504), i will add it.

christarazi reviewed (approved):
LGTM on the ENI side, pending Joe's comments.

wu0407 committed:
eni: refactor(metrics) IPAM related metrics

1. refactor ipam allocation and release metrics.
1.1 add cilium_operator_allocation_duration_seconds and cilium_operator_release_duration_secon
2. rename metrics name
2.1 `cilium_operator_ipam_available` to `cilium_operator_ipam_available_interfaces`
2.2 `cilium_operator_ipam_allocation_ops` to `cilium_operator_ipam_ip_allocation_ops`
2.3 `cilium_operator_ipam_release_ops` to `cilium_operator_ipam_ip_release_ops`

Fixes: #20704

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>
